### PR TITLE
Chameleon clothing refactor + new UI

### PR DESCRIPTION
**Pull request by Absolucy**
## About The Pull Request

This completely refactors chameleon clothing, reworking them into a component, and adding a brand new tgui-based interface for them, which allows you to actually VISUALIZE what you're disguising as!

## Why It's Good For The Game

Because componentizing good, and more usable UIs are also good.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

</details>

## Changelog
:cl:
refactor: Chameleon clothes now use components.
add: Added a brand new visual-based UI for chameleon clothing!
/:cl:

### CODE REVIEW

File: code/datums/components/chameleon/action.dm
Comment (Absolucy) [inline code/datums/components/chameleon/action.dm]: The reason this isn't part of the spritesheet like the disguises, is because for some reason, job outfit icons _do not properly insert into a spritesheet_, resulting in around half of the job icons being blank.
I have absolutely no clue why that happens, but it likely has to do with the cryptic "bad icon operation" runtimes that occur during icon generation.

**Comment by Absolucy** [comment]
Note: maintainers are free to go ahead and review this, the only reason it's a draft PR is because I need to implement the preset tab on the UI and also do some finishing touches on the UI

File: code/datums/components/chameleon/_chameleon.dm
Comment (Absolucy) [inline code/datums/components/chameleon/_chameleon.dm]:
```suggestion
	var/anyone_can_use = FALSE
```
Just reminding myself to ensure I change this back to `FALSE` before it's merged.
Not right now though, having it `TRUE` is good for testing.

**Comment by imsynthex** [comment]
List mode in the chameleon editor should be more compact, it's also quite confusing at first glance and took me a solid minute to figure out what was going on

**Comment by EvilDragonfiend** [comment]
image asset refactor PR is not merged yet and it's pain to see someone adding a new image asset

**Comment by itsmeow** [comment]
The duplicate backpacks are confusing, one PER icon state please. Also the image category buttons need captions in bold probably above the image

**Comment by Absolucy** [comment]
@itsmeow Wait, what do you mean captions?

**Comment by itsmeow** [comment]
My main gripe is still that the buttons have no text above/below them labeling what they are, and instead rely on tooltips.

When not in list view, the text should still be visible. This is especially important for the category buttons on the left.

**Comment by itsmeow** [comment]
Search bar should be wider and imo looks better with a magnifying icon like in orbit panel

**Comment by Absolucy** [comment]
I'm trying to think how I could add the text without making it look bad...

**Comment by Absolucy** [comment]
bigger search bar

**Comment by itsmeow** [comment]
The text is slightly off center oh god

**Comment by Absolucy** [comment]
@itsmeow agh i wish tgui had inspect element so i could actually see why something is off-center instead of changing stuff and guessing

File: code/datums/components/chameleon/_chameleon.dm
Comment (itsmeow) [inline code/datums/components/chameleon/_chameleon.dm]: use CRASH()

File: code/datums/components/chameleon/_chameleon.dm
Comment (itsmeow) [inline code/datums/components/chameleon/_chameleon.dm]: confused by this glob here. why is this global picking paths from a random proc

File: code/datums/components/chameleon/action.dm
Comment (itsmeow) [inline code/datums/components/chameleon/action.dm]: honestly might be best to sort by typepath of the item - it would make a lot more sense, duffel bags will show in order etc

File: code/datums/components/chameleon/spritesheet.dm
Comment (itsmeow) [inline code/datums/components/chameleon/spritesheet.dm:21]: Bllleeeeeeeeeeeeeeeeech

File: code/datums/components/chameleon/subtypes.dm
Comment (itsmeow) [inline code/datums/components/chameleon/subtypes.dm:92]: fugitive IDs (/bounty subtype) I know are not restricted here. Shouldn't this just be like a whitelist for specific subtypes?

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by itsmeow** [comment]
Updated screenshots when you're done would be epic. merge master too so you get those sweet generic theme improvements

File: code/datums/components/chameleon/_chameleon.dm
Comment (Absolucy) [reply code/datums/components/chameleon/_chameleon.dm]: what do you mean?

File: code/datums/components/chameleon/spritesheet.dm
Comment (Absolucy) [reply code/datums/components/chameleon/spritesheet.dm:21]: i can't think of a better way besides hardcoding

File: code/datums/components/chameleon/_chameleon.dm
Comment (Absolucy) [reply code/datums/components/chameleon/_chameleon.dm]: oh i see. i uh, forgot what this glob was for lol.

File: code/datums/components/chameleon/spritesheet.dm
Comment (itsmeow) [reply code/datums/components/chameleon/spritesheet.dm:21]: How was it done before

if it was hardcoded before tbh just do that. There's not really an insane amount of chameleon items

Relying on typepaths is suuuper iffy behavior

Comment (Absolucy) [reply code/datums/components/chameleon/spritesheet.dm:21]: > How was it done before

It... _wasn't_ doing this before. This is in _spritesheet generation code_. There were no sprites to generate before.

File: code/datums/components/chameleon/action.dm
Comment (EvilDragonfiend) [reply code/datums/components/chameleon/action.dm]: What was your original code which did that thing?

**Comment by EvilDragonfiend** [comment]
image

I didn't get the answer yet

File: code/datums/components/chameleon/action.dm
Comment (Absolucy) [reply code/datums/components/chameleon/action.dm]: Something like this:
```
/proc/generate_human_outfit_icon(outfit_path)
    var/mob/living/carbon/human/dummy/mannequin = generate_or_wait_for_human_dummy(DUMMY_HUMAN_SLOT_OUTFIT)
    for(var/I in mannequin.get_equipped_items(TRUE))
        qdel(I)
    mannequin.set_species(/datum/species/human, TRUE)
    mannequin.setDir(SOUTH)
    mannequin.equipOutfit(outfit_path, TRUE)
    COMPILE_OVERLAYS(mannequin)
    var/icon/asset = getFlatIcon(mannequin, no_anim = TRUE)
    asset.Scale(asset.Width() * 2, asset.Height() * 2)
    . = asset
    unset_busy_human_dummy(DUMMY_HUMAN_SLOT_OUTFIT)
```
(I don't have the full actual code, I had to dig this code snippet out of coderbus since it doesn't seem to be in the revision history anymore)

**Comment by Absolucy** [comment]
@EvilDragonfiend Wait, did I ever give you the answer?

```
/proc/generate_human_outfit_icon(outfit_path)
    var/mob/living/carbon/human/dummy/mannequin = generate_or_wait_for_human_dummy(DUMMY_HUMAN_SLOT_OUTFIT)
    for(var/I in mannequin.get_equipped_items(TRUE))
        qdel(I)
    mannequin.set_species(/datum/species/human, TRUE)
    mannequin.setDir(SOUTH)
    mannequin.equipOutfit(outfit_path, TRUE)
    COMPILE_OVERLAYS(mannequin)
    var/icon/asset = getFlatIcon(mannequin, no_anim = TRUE)
    asset.Scale(48, 48)
    . = asset
    unset_busy_human_dummy(DUMMY_HUMAN_SLOT_OUTFIT)
```

**Comment by Absolucy** [comment]
yeah il;l fix the stlopag stujff later

File: code/datums/components/chameleon/spritesheet.dm
Comment (EvilDragonfiend) [inline code/datums/components/chameleon/spritesheet.dm]:
```suggestion
		var/datum/component/chameleon/chameleon = New chameleon_path()
		// snip
		list_chameleon_disguises(..., chameleon.disguise_whitelist, ...)
		qdel(chameleon)
```
You can't use `initial()` to a var/list. You need to create a type to get a list, then del it later.

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]

**Comment by CydiaLamiales** [comment]
> balance: Only traitors/nukies/incursion members know how to use chameleon clothes by default, however they can create manuals to teach others how to use them!

Is this still necessary with multitool locking?

**Comment by Absolucy** [comment]
@CydiaLamiales good point.

**Review by PowerfulBacon — COMMENTED** [review]
This huge amount of sprites in the spritesheet and generation of human icons is going to absolutely destroy roundstart performance. The human outfit icons need to be cached and this will probably start pushing us to the point of needing reworks to assets to either cache on development or not include certain assets on development (or just have smarter caching in general) as the more icons we add as spritesheets, the longer the lag period at the start of the round and init times will be. The init times for the tests in this PR are over 90 seconds compared to other PRs which are around 84

File: code/datums/components/chameleon/action.dm
Comment (PowerfulBacon) [inline code/datums/components/chameleon/action.dm:92]: Really, we should be doing the sorting on the client side

File: code/datums/components/chameleon/spritesheet.dm
Comment (PowerfulBacon) [inline code/datums/components/chameleon/spritesheet.dm:21]: use ispath() at least

File: code/datums/components/chameleon/spritesheet.dm
Comment (PowerfulBacon) [inline code/datums/components/chameleon/spritesheet.dm:1]: I can't remember exactly the correct way to do spritesheets without looking at them, but this seems wrong. This adds all of the images in 1 go which will just completely pause until it completes compared to regular spritesheets which add all of the items they want to process and then add them one at the time. Is this really how other spritesheets are doing this, since rather than being laggy for some time it will just simply freeze the game until it completes

File: code/datums/components/chameleon/spritesheet.dm
Comment (PowerfulBacon) [inline code/datums/components/chameleon/spritesheet.dm:42]: Why scale the icons up? Scale is a pretty expensive proc here

File: code/modules/mob/inventory.dm
Comment (PowerfulBacon) [inline code/modules/mob/inventory.dm:164]: Instead of going through all slots, go through all bitflags which the item can be equipped to.

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]

File: code/datums/components/chameleon/spritesheet.dm
Comment (Absolucy) [reply code/datums/components/chameleon/spritesheet.dm:21]: This was just because there were multiple _separate_ subtypes that ended up with `/chameleon` and `/broken`, and I didn't want to hardcode those types... and I don't think `ispath` can check only the last part of a path?

**Pull request closed without merging**
